### ISSUE TITLE
[feature request] [discussion] Option to skip random weight initialization at module instance creation

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#29523]
In the typical scenario, a model is created on CPU (and the weights are initialized at CPU), then copied to GPU, then optionally `load_state_dict` is called and the weights are overwritten.

If a model is large, then generating many random numbers can take considerable time and sometimes is not necessary if `load_state_dict` is used just after.

This issue is not crucial, but if initialization argument [is introduced to module constructors](https://github.com/pytorch/pytorch/pull/1802) or [LazyTensors](https://github.com/pytorch/pytorch/issues/28341) are implemented, this can be a usecase.

A workaround is some monkey-patching of `reset_parameters` of the used modules.

One way is adding a default `reset_parameters=True` to module constructor (setting `reset_parameters=False` would skip initialization and probably even skip RAM page allocation if a new empty array is initialized on GPU instead of copying uninitialized memory, although it would break the copy invariant, but maybe it's not important to preserver the copy invariant for empty uninitialized memory)

cc @ezyang @gchanan @zou3519 @albanD @mruberry

**Comment by ezyang** [comment]
Triage review seemed like this was a reasonable thing to support. @albanD is a little nervous about having to add the parameter everywhere (context manager, maybe?)

**Comment by vadimkantorov** [comment]
Many potential ways to do this, I hope a good, conceptually simple design can be find (lazy tensors? reset_parameters argument to constructor? weight tensors arguments to constructor? state_dict/tensor_lookup argument to constructor? initializer config argument to constructor? context manager for stubbing rand factory?).

**Comment by albanD** [comment]
Here are the different designs I can think of.
Let me know what you think.

- One limitation I see is that the current `nn.Module` does not take any argument. And so all the user code calls it with no argument (not forwarding anything from their subclass to the parent class).
This means that adding a new argument is going to be tricky and potentially quite confusing to the user as it won't be a universal `nn.Module` argument.
- A context manager that controls whether or not we do the initialization could be used, but since there is no common api to initialize the Parameters, all user code need to change, otherwise it will be confusing to have some `nn.Module` affected and not others.
- A context manager that makes all the inplace random calls no-op could be done. But it feels a bit hacky and might break if users use out of place ops like `torch.rand()`.
- (My personal preference) A context manager that disables `torch.nn.init` can be done and should have a clear significance to the user. Note that it will only effect the initializations done through it. But that should be clear if we have a name like `with torch.nn.init.disabled():`. Most of the internal modules should already be using it. 
- A feasible (but O(n) in the number of nn.Module we want to support) solution would be to add an `intialize_weight=True` flag to the `nn.Module`s that are important. But we need to add that to all such modules.

**Comment by vadimkantorov** [comment]
@albanD to assess the amount of work for "new module constructor argument" scenarios, what are standard torch modules that hold learnable weights/biases:
* Conv*d
* BatchNorm*d
* Linear
* Bilinear
* Embedding
* RNN / LSTM / GRU
* MultiheadAttention

Are there more?

**Comment by albanD** [comment]
`Transformer` have some as well I think.

The more I look at it and the more I feel like disabling `torch.nn.init` is the way to go.
It is used everywhere internally.
By design, the api is only inplace ops.
It will work with user code where they use it.
The contest manager will be clear enough to avoid confusion on where it works/doesn't work.
Does not require to pass an argument to every single module you create.

**Comment by vadimkantorov** [comment]
Some arguments against context manager: inherited spreading effect, when dealing with code non-directly called by the user. Practically it's probably OK. When tracing / jit'ing, would such global flag affecting behavior change be unclear to save? 

The original [initializer suggestion](https://github.com/pytorch/pytorch/pull/18020) is also practically useful and Keras/TF have [this](https://keras.io/initializers/). Often people would have their own init-functions instead of more laconic initializer argument (esp. when complaining that PyTorch uses not fully correct Kaiming He's Conv/Linear init), and this would be probably a notable (small, but still) feature-parity thing.

I think Transformer is just a collection of TransformerEncoder/Decoder layers that use MultiheadAttention under the hood, but maybe you are right. If these are all places, it's not too much and users then could modify the code to forward this extra arg if they want to use of this feature

**Comment by vadimkantorov** [comment]
I also wonder if one could do a BigQuery search to find projects using PyTorch that do some manual initialization with nn.init. Maybe looking at code snippets would also provide some food for design

**Comment by apaszke** [comment]
I second the opinion that adding arguments is a bad idea, primarily because this approach doesn't scale. It's an easy thing to do for "leaf" modules like Linear or Conv2d, but I don't think it's reasonable to e.g. expect authors of standard implementations of models like ResNet to provide an argument to override the initialization of each single part. On the other hand, you haven't solved the problem at all unless they have.

Context managers also don't seem like a good choice IMO. There are many different kinds of parameters in those models (weights, biases, but who knows what will those be in the future), and so there's no clear way to say how would you pick between the different initialization schemes available in your context.

I don't really have a good idea for how to solve this problem, but I wanted to point out that both of those approaches will not compose well, and hence I don't think we should include them.

**Comment by apaszke** [comment]
Also, having a context manager that "disables" `torch.nn.init` feels like a bad idea to me. It's very specific to this particular case, and we'll likely run into similar issues in the future. If someone _really_ wants this, then they can monkey patch all of those methods. Finally, what if someone hard-codes the initialization and doesn't use `torch.nn.init` at all?

**Comment by vadimkantorov** [comment]
Another take: if PyTorch decides to provide initializer argument for other, usability reasons, then suppporting this is easy.

@apaszke I think that if the there is a recommended supported way of doing this, then standard model authors for big models (for which this can add a few seconds of startup time) would update their code (for instance vision models in torchvision already have an argument related to initialization in their factory methods: download weights from model hub or not), especially would use it for newer code. Also a plus is that this argument will be not bc-breaking.

**Comment by vadimkantorov** [comment]
a little bit related: https://github.com/pytorch/pytorch/pull/31143

**Comment by vadimkantorov** [comment]
it seems that caffe2 python codebase chose to have weight_init/bias_init, caffe2 people then may have some feedback about this design choice

**Comment by vadimkantorov** [comment]
This is even more true for giant pretrained models. Generating hundreds of millions random numbers must be quite considerable, given that some `from_pretrained` would reset them

> Also, having a context manager that "disables" `torch.nn.init` feels like a bad idea to me. It's very specific to this particular case, and we'll likely run into similar issues in the future. If someone _really_ wants this, then they can monkey patch all of those methods. Finally, what if someone hard-codes the initialization and doesn't use `torch.nn.init` at all (because e.g. it doesn't provide the one they're looking for)?

@apaszke @albanD One way first could be ensuring that all core `pytorch`/`fairseq`/`torchvision`... modules use `torch.nn.init_` for all initialization. And then one could check if patching out nn.init this in some fairseq pretrained bert loading brings the expected speed up. If it does, then maybe torch.nn.init.disable context manager (or a snippet recommendation in docs) is not so bad

**Comment by ezyang** [comment]
Bumping priority so we don't forget about this

**Comment by vadimkantorov** [comment]
One other useful pattern could also be to preferably do initialization only in sth like `reset_parameters/weights()` (already done in some modules) - then one could still reset initialization if needed (can be helpful in other contexts too).

One problem with this is not very clear semantics for composite/container models: should it recursively call reset weights? But for non-containers it's quite useful.

**Comment by stas00** [comment]
FWIW, we have the same issue in https://github.com/huggingface/transformers/issues/9205

**Comment by jbschlosser** [comment]
This is now supported for all modules provided by `torch.nn`. Example:

```python
import torch

# Skip initialization by instantiating on the meta device.
m = torch.nn.Linear(10, 1, device='meta')

# Materialize on CPU with uninitialized storage.
m.to_empty(device='cpu')

# Do custom init.
with torch.no_grad():
    torch.nn.init.xavier_uniform_(m.weight)
```

**Comment by vadimkantorov** [comment]
Would this be supported e.g. for torchvision models? So that one could directly load a model onto a GPU without weight init, and then load_state_dict into it directly without any random init?

If yes, it should probably be documented in a recipe, and maybe even become default for the from_pretrained-like factories

**Comment by vadimkantorov** [comment]
Was `reset_parameters = True` argument decided finally not taken? Should I create a separate issue about standardizing practice of putting all (standard models') weight init into `reset_parameters` method (so that CPU-side weight init could be skipped if needed and be done on GPU directly if needed)

**Comment by ezyang** [comment]
@jbschlosser I think we need to document this recipe. Also, weren't we going to add a helper function to help you do this two step in one line? It also seems reasonable to start applying this pattern when doing stuff like `from_pretrained`.

@vadimkantorov We dumped `reset_parameters` because it turned out that `device="meta"` and then `to_empty` was sufficient, with less refactoring required on the Module side.

**Comment by vadimkantorov** [comment]
Even if it's functionally equivalent, I must admit that `devcie="meta"` and `to_empty` is quite cryptic from the end-user standpoint

**Comment by jbschlosser** [comment]
Fair enough, I shouldn't have closed this prematurely :)

> Would this be supported e.g. for torchvision models? So that one could directly load a model onto a GPU without weight init (on CPU or on GPU), and then load_state_dict into it directly without any random init?

Modules under `torch.nn` now support this, so anything composed of those modules can follow the same pattern. While there's nothing preventing this from working with torchvision models, they haven't been updated yet to accept a device arg and string it along through the submodules.

> @jbschlosser I think we need to document this recipe. Also, weren't we going to add a helper function to help you do this two step in one line? It also seems reasonable to start applying this pattern when doing stuff like from_pretrained.

For sure- I'll be opening a docs PR shortly adding this recipe. I'll add the helper function as well and agree it'd be good to use it in the various `from_pretrained()` functions. For the case of the helper function being called on a module without the `device` arg, I intend to throw an informative error describing this caveat and explaining the opt-in module modification needed for the support.

> Even if it's functionally equivalent, I must admit that `devcie="meta"` and `to_empty` is quite cryptic from the end-user standpoint (to just omit random weight init). OTOH, it's good it became possible at all :)

Yeah, I don't disagree with you here :) The helper function @ezyang mentioned is intended to help out the UX. As mentioned above, it will only work for modules that accept a `device` arg due to the implementation details. The current work technically allows for skipping random init or direct init onto non-CPU devices and, while it isn't the most intuitive UX yet, the idea was to support those use cases for those who really need them.

> Was reset_parameters = True argument decided finally not taken? Should I create a separate issue about standardizing practice of putting all (standard models') weight init into reset_parameters method (so that CPU-side weight init could be skipped if needed and be done on GPU directly if needed)

Supporting `reset_parameters=True` (or alternatively, `init_parameters=True`) in module constructors wasn't ruled out entirely, but was conservatively left out of the initial work. There's at least one weird kink that comes up when trying to add a `reset_parameters()` method to all modules: since a subclass implementation of this method will override the base implementation, an error will be thrown for trying to initialize a not-yet-created derived class parameter from the base constructor during the call to `super()`. While this can be worked around by e.g. passing `init_parameters=False` in the call to `super()` or explicitly qualifying `BaseClass.reset_parameters()`, the results are unintuitive, so we're unsure if we want to take this on as part of the module standard quite yet.

**Comment by vadimkantorov** [comment]
Also, currently reset_parameters methods seem not documented: https://pytorch.org/docs/stable/search.html?q=reset_parameters&check_keywords=yes&area=default

So changing the name now may be a good time.

**Comment by jbschlosser** [comment]
The helper function `torch.nn.utils.skip_init()` was added, and instructions on using it and updating modules to support it are [here](https://pytorch.org/tutorials/prototype/skip_param_init.html).

**Comment by vadimkantorov** [comment]
Is there some public tutorial on using this feature for users in some realistic context? And same for adoption in torchvision... Should I create a new separate issue for this?

**Comment by vadimkantorov** [comment]
Maybe worth to advertise to @stas00 with HuggingFace people?

**Comment by jbschlosser** [comment]
> Maybe worth to advertise to @stas00 with HuggingFace people?

Great suggestion! I agree

**Comment by stas00** [comment]
Thank you for the heads up, @vadimkantorov!

We meanwhile implemented our own solution: https://github.com/huggingface/transformers/pull/11471

and also we won't be able to use this solution for a while since we have to remain back-compatible to older pytorch which doesn't have this new feature. But definitely something to switch to in the future!

**Comment by vadimkantorov** [comment]
Also, maybe some conventions are needed on automatic calls of `reset_parameters` that defeat the meta tensors: https://github.com/pytorch/vision/pull/4403/files#diff-62577d15011d998701098e5e567a78bd117257deeeb2a23b72e550e1f6b14279R559

**Comment by vadimkantorov** [comment]
Another perf improvement could be a load_state_dict + lazy torch.load impl. It would avoid having to read all the weights on RAM at once and would directly allow to copy the pretrained weights onto GPU (if the model is going to be in GPU at the end). Maybe the most impactful and self-contained could be torch.load returning LazyTensors that are actually read from file when the tensor contents is requested. Then existing load_state_dict would do fine. One could also imagine some sort of memory-mapped formats where one can directly copy from disk file onto GPU with chunked copy.

Related ideas: https://github.com/pytorch/pytorch/issues/64601 https://github.com/pytorch/pytorch/issues/76441 https://github.com/pytorch/pytorch/issues/64625

**Comment by stas00** [comment]
That would be amazing, @vadimkantorov!

I'm trying to get Deepspeed to implement that, but if pytorch would have that it'd be wonderful, as it'd be available under any framework.

I have made various proposals/requests:
1. https://github.com/pytorch/pytorch/issues/64327
2. https://github.com/pytorch/pytorch/issues/75242

in particular the first one where the `state_dict` is a tied dbm which of course gets automatically read/written to/from disc as needed and this is done transparently to the user.

**Comment by vadimkantorov** [comment]
I wonder if some sort of lazytensor could be a good concept for implementing this…

**Comment by albanD** [comment]
> I wonder if some sort of lazytensor could be a good concept for implementing this…

Note that now that you have Tensor subclass, you can relatively easily create such a class from python.
That might require pre-processing the checkpoint though to make it possible to read each Tensor independently.

**Comment by vadimkantorov** [comment]
I guess for load_state_dict to work with such a subclass, it should support some interface so that torch.as_tensor would work on it. Also, it would be nice for load_state_dict to call torch.as_tensor on values to be loaded. Then even NumPy / HDF5+h5py dataset objects should work https://github.com/pytorch/pytorch/issues/67549

**Comment by stas00** [comment]
and one of the performance-oriented ideas is to deploy NVME drives with GPUDirect https://developer.nvidia.com/gpudirect, which allows one to directly move data GPU <=> NVME which would be (much?) faster than doing it via CPU.

**Comment by vadimkantorov** [comment]
@albanD btw load_state_dict could check and call `copyto` in some cases to transform param.copy_(lazy_tensor) to lazy_tensor.copyto(param) that would do smarter chunk-based or mmap or GPUDirect copy.

Btw in NumPy exist a functional method numpy.copyto(src, dst), so introducing an instance method lazy_tensor.copyto would make perfect sense, and then torch.copyto(src, dst) could also be implemented for NumPy compat

**Comment by vadimkantorov** [comment]
i guess the benefit could be if the object-to-be-copied from is a dynamic python-defined type (that has copyto method) that the copy_ dispatched knows nothing of

**Comment by albanD** [comment]
That's where `__torch_function__` shines, it actually make all of the `torch.` function dynamically dispatched! https://pytorch.org/docs/master/notes/extending.html#extending-torch has more details

**Comment by vadimkantorov** [comment]
@albanD I factored out some of this discussion into https://github.com/pytorch/pytorch/issues/79967

**Comment by vadimkantorov** [comment]
@albanD it seems that your proposed torch_dispatch got used by @t-vi @stas00

https://mobile.twitter.com/StasBekman/status/1632064037788979203
http://lernapparat.de/faster-model-init

Should this also become built-in?
